File: functions/src/game/deck.ts
export const Suits = ['clubs', 'diamonds', 'hearts', 'spades'] as const;
export type Suit = typeof Suits[number];

export const Ranks = ['two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', 'jack', 'queen', 'king', 'ace'] as const;
export type Rank = typeof Ranks[number];

export interface Card {
  suit: Suit;
  rank: Rank;
}

export function createDeck(): Card[] {
  const deck: Card[] = [];
  for (const suit of Suits) {
    for (const rank of Ranks) {
      deck.push({ suit, rank });
    }
  }
  return deck;
}

export function shuffle<T>(array: T[]): T[] {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

File: functions/lib/game/deck.dart
enum Suit { clubs, diamonds, hearts, spades }

enum Rank {
  two,
  three,
  four,
  five,
  six,
  seven,
  eight,
  nine,
  ten,
  jack,
  queen,
  king,
  ace
}

class Card {
  final Suit suit;
  final Rank rank;

  Card({required this.suit, required this.rank});

  Map<String, dynamic> toJson() => {
        'suit': suit.name,
        'rank': rank.name,
      };

  factory Card.fromJson(Map<String, dynamic> json) => Card(
        suit: Suit.values.byName(json['suit'] as String),
        rank: Rank.values.byName(json['rank'] as String),
      );

  @override
  String toString() => '${rank.name} of ${suit.name}';
}

List<Card> createDeck() {
  final deck = <Card>[];
  for (final suit in Suit.values) {
    for (final rank in Rank.values) {
      deck.add(Card(suit: suit, rank: rank));
    }
  }
  return deck;
}

List<T> shuffle<T>(List<T> array) {
  final shuffled = List<T>.from(array);
  shuffled.shuffle();
  return shuffled;
}
